torch/_functorch/_aot_autograd/fx_passes.py

```diff
@@ -0,0 +1,152 @@
+from collections import defaultdict
+from typing import Dict, List
+
+import torch
+from torch import fx
+
+
+def is_primal(node: fx.Node) -> bool:
+    return node.op == "placeholder" and "tangents" not in str(node.target)
+
+
+def move_resize_zero_to_end_of_graph(graph: fx.Graph) -> None:
+    node_list = list(graph.nodes)
+    return_node = node_list[-1]
+    assert return_node.target == "output"
+    for n in node_list:
+        if (
+            n.op == "call_function"
+            and n.target is torch.ops.inductor.resize_storage_bytes_.default
+            and n.args[1] == 0
+        ):
+            resize_node = n
+            with graph.inserting_before(return_node):
+                new_resize_node = graph.call_function(
+                    torch.ops.inductor.resize_storage_bytes_.default, resize_node.args
+                )
+            graph.erase_node(resize_node)
```

Comment:
Hmm, we probably want some more safety and invariant checks to happen if we're going to do this:

My understanding is: the invariant in FSDP is that after we call `param.resize_(0)`, `param` is guaranteed to be unused in the remainder of the graph, which is what makes moving the resize_() to the end "trivial". If that's true, can we assert that there are no uses of `n.args[0]` anywhere in the graph after this node?

Reply:
yes agreed - we should assert that there are no uses of `n.args[0]` anywhere in the graph after this node